src/pudl/extract/excel.py

```diff
@@ -354,3 +363,103 @@ def excel_filename(self, page, **partition):
             string name of the xlsx file
         """
         return self.METADATA.get_file_name(page, **partition)
+
+
+@op
+def merge_dfs_by_page(
+    paged_dfs: list[dict[str, pd.DataFrame]]
+) -> dict[str, pd.DataFrame]:
+    """Merge DataFrames loaded with independent calls to .extract().
+
+    The input is a list of dictionaries where the keys are the page names,
+    and values are DataFrame of extracted data.
+    Where a page name appears more than once in the dictionaries, all
+    DataFrames with that page name are merged together, and the result
+    returned as a single dict mapping unique page names to merged
+    DataFrames.
+    """
+    merged = defaultdict(list)
+    for dfs in paged_dfs:
+        for page in dfs:
+            merged[page].append(dfs[page])
+
+    for page in merged:
+        merged[page] = pd.concat(merged[page]).reset_index(drop=True)
+
+    return merged
+
+
+def year_loader_factory(
+    extractor_cls: type[GenericExtractor], name: str
+) -> OpDefinition:
+    """Return a dagster op to load a single year using a given extractor class.
+
+    Args:
+        extractor_cls: Class of type GenericExtractor used to extract the data.
+        name(str): Name of dataset (e.g. "eia860").
+    """
+
+    def load_single_year(context, year: int) -> dict[str, pd.DataFrame]:
+        """Load a single year of EIA data from file.
+
+        Args:
+            context:
+                context: dagster keyword that provides access to resources and config.
+            year:
+                Year to load.
+
+        Returns:
+            Loaded data in a dataframe.
+        """
+        ds = context.resources.datastore
+        return extractor_cls(ds).extract(year=[year])
+
+    return op(
+        required_resource_keys={"datastore", "dataset_settings"},
+        name=f"load_single_{name}_year",
+    )(load_single_year)
+
+
+def years_from_settings_factory(name: str) -> OpDefinition:
+    """Return a dagster op to get EIA years from settings.
+
+    Args:
+        name (str): Name of dataset (e.g. "eia860")
+    """
+
+    def years_from_settings(context):
```

Comment:
How does `years_from_settings()` end up having access to `context`?

Reply:
I think Dagster just conjures it for all ops: https://docs.dagster.io/concepts/ops-jobs-graphs/ops#op-context